src/Common/FST.cpp

```diff
@@ -0,0 +1,445 @@
+#include <cassert>
+#include <iostream>
+#include <memory>
+#include <vector>
+#include <algorithm>
+#include <../contrib/consistent-hashing/popcount.h>
+#include <IO/WriteHelpers.h>
+#include <IO/ReadHelpers.h>
+#include <Common/HashTable/Hash.h>
+#include <Common/Exception.h>
+#include "FST.h"
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+};
+
+UInt16 Arc::serialize(WriteBuffer& write_buffer) const
+{
+    UInt16 written_bytes = 0;
+    bool has_output = output != 0;
+
+    /// First UInt64 is target_index << 1 + has_output
+    UInt64 first = ((target->state_index) << 1) + has_output;
+    writeVarUInt(first, write_buffer);
+    written_bytes += getLengthOfVarUInt(first);
+
+    /// Second UInt64 is output (optional based on whether has_output is not zero)
+    if (has_output)
+    {
+        writeVarUInt(output, write_buffer);
+        written_bytes += getLengthOfVarUInt(output);
+    }
+    return written_bytes;
+}
+
+void ArcsBitmap::addArc(char label)
+{
+    uint8_t index = label;
+    UInt256 bit_label = 1;
+    bit_label <<= index;
+
+    data |= bit_label;
+}
+
+int ArcsBitmap::getIndex(char label) const
+{
+    int bit_count = 0;
+
+    uint8_t index = label;
+    int which_int64 = 0;
+    while (true)
+    {
+        if (index < 64)
+        {
+            UInt64 mask = index == 63 ? (-1) : (1ULL << (index+1)) - 1;
+
+            bit_count += PopCountImpl(mask & data.items[which_int64]);
+            break;
+        }
+        index -= 64;
+        bit_count += PopCountImpl(data.items[which_int64]);
+
+        which_int64++;
+    }
+    return bit_count;
+}
+
+int ArcsBitmap::getArcNum() const
+{
+    int bit_count = 0;
+    for (uint64_t item : data.items)
+    {
+        if (item)
+            bit_count += PopCountImpl(item);
+    }
+    return bit_count;
+}
+
+bool ArcsBitmap::hasArc(char label) const
+{
+    uint8_t index = label;
+    UInt256 bit_label = 1;
+    bit_label <<= index;
+
+    return ((data & bit_label) != 0);
+}
+
+Arc* State::getArc(char label)
+{
+    auto it(arcs.find(label));
+    if (it == arcs.cend())
+        return nullptr;
+
+    return &it->second;
+}
+
+void State::addArc(char label, Output output, StatePtr target)
+{
+    arcs[label] = Arc(output, target);
```

Comment:
I double checked the algorithm and overwriting case won't happen. Since it is called intensively and we don't want to slow it down even in debug mode by assert(), so let's leave it as is IMHO.